### ISSUE TITLE
Issue with quantity precision

### ISSUE THREAD

**Issue by Swekkie** [Binance-docs/Binance_Futures_python#16]
Hi

Came across the issue by:
`request_client.post_order(symbol="ETHUSDT", side=OrderSide.BUY, ordertype=OrderType.LIMIT, timeInForce="GTC", quantity=0.015, price=222)`

leading to the following error:

> binance_f.exception.binanceapiexception.BinanceApiException: ('ExecuteError', '[Executing] -1111: Precision is over the maximum defined for this asset.')

This is the URL:

> url:/fapi/v1/order?symbol=ETHUSDT&side=BUY&type=LIMIT&timeInForce=GTC&quantity=0.01499999999999&price=222 ...

So the quantity in the url should be 0.015

Problem probably will be in the "put_url" function in "urlparamsbuilder.py" when value is instance of float

If I comment out lines 15 and 16:
`            # elif isinstance(value, float):`
`          #   self.param_map[name] = ('%.20f' % (value))[slice(0, 16)].rstrip('0').rstrip('.')`
it works fine (0.015 in url) so I would think that those lines can be deleted

Greetz
Sven

**Comment by Swekkie** [comment]
You could strip 9's to but if I put quantity=0.015, I expect the order to be 0.015 and not 0.014, so I would just take the str(0.015) and put that in the url

**Comment by freemanjackal** [comment]
I faced a similar problem some times trying to make post orders...it was formatting the price and putting instead 2 decimal values 16(13 after the dot). I Didn't comment the lines but a workaround... it is working fine so far...

`self.param_map[name] = format_float(value)`

> import decimal
>  
> def format_float(value):
> &nbsp; &nbsp; &nbsp; &nbsp;ctx = decimal.Context()
> &nbsp; &nbsp; &nbsp; &nbsp;    ctx.prec = 16
> &nbsp; &nbsp; &nbsp; &nbsp;    test = ctx.create_decimal(repr(value))
> &nbsp; &nbsp; &nbsp; &nbsp;    return test

**Comment by kimchirichie** [comment]
Also facing same issues.
```
quantity = round(929.812123131, 1)
result = request_client.post_order(
          symbol="IOTAUSDT"
          side=OrderSide.BUY,
          ordertype=OrderType.MARKET,
          quantity=quantity,
)
```

yields 
`url:/fapi/v1/order?symbol=IOTAUSDT&side=BUY&type=MARKET&quantity=929.799999999999&recvWindow=60000&timestamp=1584431264109&signature=23ba252370f3a625c391526a37866b65653f0aec1153772905caf0adb66abaa8`

**Comment by funoverip** [comment]
I think this issue is valid for any library talking to binance API.

The following API should give you the **precision** and **step** for each symbol: https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#exchange-information

Then the following filters should be applied to your quantity/price:
https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#symbol-filters

And at the end, you'll have to format your quantity/price **precision** according to the symbol details (as string), example using :


```
amount = 0.000234234
precision = 5
amt_str = "{:0.0{}f}".format(amount, precision)
```

Example coming from : https://python-binance.readthedocs.io/en/latest/account.html#orders

**Comment by xinqiyang** [comment]
> Also facing same issues.
> 
> ```
> quantity = round(929.812123131, 1)
> result = request_client.post_order(
>           symbol="IOTAUSDT"
>           side=OrderSide.BUY,
>           ordertype=OrderType.MARKET,
>           quantity=quantity,
> )
> ```
> 
> yields
> `url:/fapi/v1/order?symbol=IOTAUSDT&side=BUY&type=MARKET&quantity=929.799999999999&recvWindow=60000&timestamp=1584431264109&signature=23ba252370f3a625c391526a37866b65653f0aec1153772905caf0adb66abaa8`

this is ok , I fix it , thank you @kimchirichie

**Comment by kuegi** [comment]
i faced the same issue, but i don't like changing library code for sake of updates etc. so i found a "hacky" way around without changing the lib code: it defines float as type for the quantity, but never actually cares about it.
so if you format the quantity into a string yourself (with the precision and all) and pass that string as argument to the post_order method, your IDE might warn you. but the code runs smoothly, adding the string to the url instead of misformating the float.

**Comment by D50000** [comment]
This issue can easy work normally by pass in the ```price = "200"```. (not **float** is actually a **string**)
I think the float input in the example mean that the class in the base implement declare in float. But when https request to the server is passing the string.

**Comment by kuegi** [comment]
> This issue can easy work normally by pass in the `price = "200"`. (not **float** is actually a **string**)
yes you can ignore the definition of the function and just pass in a str. But good IDE starts to warn you, and the API should declare the parameter a float if you need to pass a str to prevent errors.

signature of the method is `def post_order(self, symbol: 'str', side: 'OrderSide', ordertype: 'OrderType', 
                timeInForce: 'TimeInForce' = TimeInForce.INVALID, quantity: 'float' = None,
                reduceOnly: 'boolean' = None, price: 'float' = None,
                newClientOrderId: 'str' = None, stopPrice: 'float' = None, 
                workingType: 'WorkingType' = WorkingType.INVALID) -> any:`

**Comment by kimchirichie** [comment]
for those who are scrolling for quick answer:
```
quantity = round(929.812123131, 1)
result = request_client.post_order(
  symbol="IOTAUSDT"
  side=OrderSide.BUY,
  ordertype=OrderType.MARKET,
  quantity=str(quantity), # must be string because the lib will append wrong precision
)
```

**Comment by MachineTrader** [comment]
The library says it want a float, shouldn't the library take care of the converting for us then?

**Comment by OkayQuest** [comment]
I edited the lines 15 - 17 in the  "urlparamsbuilder.py" found in " \binance_f\impl\utils" as follows:
            elif isinstance(value, float):
                value_str = str(value)
                self.param_map[name] = value_str

The code self.param_map[name] = ('%.20f' % (value))[slice(0, 16)].rstrip('0').rstrip('.') is practically formatting the float to string with some string operations. The line  I replaced it with (self.param_map[name] = value_str) helps me achieve similiar objective but without the precision error.

Ensure that you send the EXACT precision float (price, qty etc), as defined by Binance API, from your code and the library will just convert it exactly as it is to string and parse to the url.

I believe it should also work for binance_d ( I did not try it).

**Comment by teranoib** [comment]
```
tradingPairs = ['BTCUSDT','ETHUSDT','BNBUSDT']

#Loop though cryptos

for i in range(0,len(tradingPairs)):

info = client.futures_exchange_info()

if info['symbols'][0]['pair'] == tradingPairs[i]:

print("Price Pre ",info['symbols'][0]['pricePrecision'])

pricePrecision = info['symbols'][0]['pricePrecision']
quantityS = 5.2
quantityB = "{:0.0{}f}".format(quantityS, pricePrecision)
```
